api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -286,87 +287,92 @@ def __init__(
+        query = exp.axis_query(
             measurement_name=self.measurement_name,
             obs_query=self.obs_query,
             var_query=self.var_query,
         )
 
-        obs_joinids = self._query.obs_joinids()
+        # The to_numpy() call is a workaround for a possible bug in TileDB-SOMA:
+        # https://github.com/single-cell-data/TileDB-SOMA/issues/1456
+        self._obs_joinids = query.obs_joinids().to_numpy()
+        self._var_joinids = query.var_joinids().to_numpy()
+
+        self._encoders = self._build_obs_encoders(query)
 
+        self._initialized = True
+
+    @staticmethod
+    def _partition_obs_joinids(ids: npt.NDArray[np.int64]) -> npt.NDArray[np.int64]:
+        # NOTE: Can alternately use a `worker_init_fn` to split among workers split workload
         worker_info = torch.utils.data.get_worker_info()
-        if worker_info:
-            # multi-processing mode
 
-            if worker_info.num_workers > 0 and self.sparse_X:
-                raise NotImplementedError(
-                    "torch does not work with sparse tensors in multi-processing mode "
-                    "(see https://github.com/pytorch/pytorch/issues/20248)"
-                )
+        if worker_info is None or worker_info.num_workers < 2:
+            return ids
 
-            partition, num_partitions = worker_info.id, worker_info.num_workers
+        partition, num_partitions = worker_info.id, worker_info.num_workers
 
-        else:
-            partition, num_partitions = 0, 1
-
-        if num_partitions is not None:
-            # partitioned data loading
-            # NOTE: Can alternately use a `worker_init_fn` to split among workers split workload
-            partition_size = len(obs_joinids) // num_partitions
-            partition_start = partition_size * partition
-            partition_end_excl = min(len(obs_joinids), partition_start + partition_size)
-            # The to_numpy() call is a workaround for a possible bug in TileDB-SOMA:
-            # https://github.com/single-cell-data/TileDB-SOMA/issues/1456
-            self._obs_joinids_partitioned = obs_joinids[partition_start:partition_end_excl].to_numpy()
-
-            if pytorch_logger.isEnabledFor(logging.DEBUG):
-                if self._obs_joinids_partitioned is not None and len(self._obs_joinids_partitioned) > 0:
-                    joinids_start = self._obs_joinids_partitioned[0]
-                    joinids_end = self._obs_joinids_partitioned[-1]
-                else:
-                    joinids_start = joinids_end = 0
-
-                pytorch_logger.debug(
-                    f"Process {os.getpid()} handling partition {partition + 1} of {num_partitions}, "
-                    f"index range={partition_start}:{partition_end_excl}, "
-                    f"soma_joinid range={joinids_start}:{joinids_end}, "
-                    f"{partition_size:}"
-                )
+        partition_size = len(ids) // num_partitions
```

Comment:
maybe just use numpy.array_split?  (up to you - only saves 4 lines of code)

Reply:
No maybes about it--done!